### PR TITLE
[WIP] Fix for missing IntelliSense after type constraint error

### PR DESCRIPTION
**Pull request by mpetruska**
work in progress...

Fixes issue #629.
---

Note: contains the changes in PR #841 .

### CODE REVIEW

**Comment by mpetruska** [comment]
Currently the missing name resolutions are fixed:
intellisense_screenshot

A type constraint error occurs when the type checker tries to unify types `unit` and `'a list`, after which type inference is not run to specialize `List.map`, hence the type of the function parameter (`abcd`) is left as `'a`.

Further work is needed to try to infer the types for IntelliSense.

File: src/fsharp/TypeChecker.fs [hunk 1]
```diff
@@ -5067,7 +5067,18 @@ and RecordNameAndTypeResolutions_IdeallyWithoutHavingOtherEffects_Delayed cenv e
...
+    dummyCheckedDelayed delayed
+
+// Calls UnifyTypes, records name resolutions on failure, so that type information
+// can be collected for IntelliSense.
+and UnifyTypes_RecordNameResolutionsOnError cenv env m expectedTy actualTy tpenv delayed =
+    try
+        UnifyTypes cenv env m expectedTy actualTy
+    with
+        _ ->
+            RecordNameAndTypeResolutions_IdeallyWithoutHavingOtherEffects_Delayed cenv env tpenv delayed
```
Comment (mpetruska) [inline src/fsharp/TypeChecker.fs]: This unfortunately hurts performance, but only in case an error occurs in `UnifyTypes`.

Comment (dsyme) [inline src/fsharp/TypeChecker.fs]: That's totally ok

**Comment by dsyme** [comment]
Interesting fix.  I believe I've seen other manifestations of this problem.  Does the change pass all the IDE nunit tests?  If you could add a new IDE unit test that would be great.

**Comment by mpetruska** [comment]
Unfortunately I'm having some trouble with the IDE unit tests.

Running the IDE tests (`RunTests.cmd debug ideunit`) without the changes gives the following results:
```
Tests run: 4146, Passed: 3289, Errors: 154, Failures: 703, Inconclusive: 0
```

With my changes the result is:
```
Tests run: 4149, Passed: 3292, Errors: 154, Failures: 703, Inconclusive: 0
```

In my opinion the type inference problem (see above) can be solved, will continue to work on it.

**Comment by enricosada** [comment]
@mpetruska what are the ideunit failures?

**Comment by mpetruska** [comment]
@enricosada 
Here is the log of the test run without the changes: [IDEUnit_Output_base.log.txt](https://github.com/Microsoft/visualfsharp/files/81896/IDEUnit_Output_base.log.txt)
Here is the one that contain my changes: [IDEUnit_Output_withchanges.log.txt](https://github.com/Microsoft/visualfsharp/files/81900/IDEUnit_Output_withchanges.log.txt)

I checked a couple of failures manually (only IntelliSense ones), they seemed to work correctly in Visual Studio 2015. I think there's some configuration issue on my box, I'm using Windows 8.1 and the Community edition of VS 2015 (Version 14.0.24720.00 Update 1, .Net 4.6).

**Comment by KevinRansom** [comment]
@mpetruska Hi,  how is this PR progressing?  should I close it, or is it ready for review?

**Comment by mpetruska** [comment]
Hi @KevinRansom, sorry for the late reply, unfortunately I could not find time to update this one. Also I think the code is so out-of-date, that it's better to close it.

**Comment by KevinRansom** [comment]
I will keep it around, and maybe someone will have the time to pick it up.

Kevin

**Comment by KevinRansom** [comment]
Closing

**Pull request closed without merging**